src/operator/numpy/np_trace_op-inl.h

```diff
@@ -0,0 +1,255 @@
+ */
+
+#ifndef MXNET_OPERATOR_NUMPY_NP_TRACE_OP_INL_H_
+#define MXNET_OPERATOR_NUMPY_NP_TRACE_OP_INL_H_
+
+#include <dmlc/parameter.h>
+#include <mxnet/operator_util.h>
+#include <vector>
+#include <utility>
+#include <algorithm>
+#include "../mxnet_op.h"
+#include "../operator_common.h"
+#include "../elemwise_op_common.h"
+#include "../tensor/broadcast_reduce_op.h"
+
+namespace mxnet {
+namespace op {
+
+struct NumpyTraceParam: public dmlc::Parameter<NumpyTraceParam> {
+    int offset, axis1, axis2;
+    DMLC_DECLARE_PARAMETER(NumpyTraceParam) {
+        DMLC_DECLARE_FIELD(offset)
+            .set_default(0)
+            .describe("Offset of the diagonal from the main diagonal. "
+                      "Can be both positive and negative. Defaults to 0.");
+        DMLC_DECLARE_FIELD(axis1)
+            .set_default(0)
+            .describe("Axes to be used as the first axis of the 2-D sub-arrays "
+                      "from which the diagonals should be taken. Defaults to 0.");
+        DMLC_DECLARE_FIELD(axis2)
+            .set_default(1)
+            .describe("Axes to be used as the second axis of the 2-D sub-arrays "
+                      "from which the diagonals should be taken. Defaults to 1.");
+    }
+};
+
+template<int ndim, int req, bool back>
+struct numpy_trace {
+  template<typename DType>
+  MSHADOW_XINLINE static void Map(index_t i, DType* out, const DType* a,
+                                  mshadow::Shape<ndim> oshape,
+                                  mshadow::Shape<ndim> ishape,
+                                  index_t stride, index_t offset, int dlength) {
+    using namespace mxnet_op;
+    using namespace mshadow;
+    index_t j = ravel(unravel(i, oshape), ishape) + offset;
+    if (back) {
+      for (index_t k = 0; k < dlength; ++k) {
+        KERNEL_ASSIGN(out[j], req, a[i]);
+        j += stride;
+      }
+    } else {
+      if (req == kWriteTo) {
+         out[i] = 0;
+        for (index_t k = 0; k < dlength; ++k) {
+          out[i] += a[j];
+          j += stride;
+        }
+      } else if (req == kAddTo) {
+        for (index_t k = 0; k < dlength; ++k) {
+          out[i] += a[j];
+          j += stride;
+        }
+      }
+    }
+  }
+};
+
+
+template<typename xpu, bool back>
+void NumpyTraceOpProcess(const TBlob& in_data,
+                   const TBlob& out_data,
+                   const mxnet::TShape& ishape,
+                   const mxnet::TShape& oshape,
+                   index_t dsize,
+                   const NumpyTraceParam& param,
+                   mxnet_op::Stream<xpu> *s,
+                   const std::vector<OpReqType>& req) {
+  using namespace mxnet_op;
+  using namespace mshadow;
+  if (dsize == 0) {
+    if (back) {
+      if (out_data.Size() != 0) {
+        MSHADOW_TYPE_SWITCH(out_data.type_flag_, DType, {
+          MXNET_ASSIGN_REQ_SWITCH(req[0], req_type, {
+            if (req_type == kWriteTo)
+              out_data.FlatTo1D<xpu, DType>(s) = 0;
```

Comment:
please do
```c++
if (...) {
  // your code here
}
```
even for 1-line if-statements.

Reply:
fixed